llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp

```diff
@@ -0,0 +1,1242 @@
+      }
+    }
+
+    return Modified;
+  }
+
+  bool IsRequiredForPhiNode(BasicBlock *BB) {
+    for (BasicBlock *Successor : successors(BB)) {
+      for (PHINode &Phi : Successor->phis()) {
+        if (Phi.getBasicBlockIndex(BB) != -1)
+          return true;
+      }
+    }
+
+    return false;
+  }
+
+  bool removeUselessBlocks(Function &F) {
+    std::vector<BasicBlock *> ToRemove;
+
+    auto MergeBlocks = getMergeBlocks(F);
+    auto ContinueBlocks = getContinueBlocks(F);
+
+    for (BasicBlock &BB : F) {
+      if (BB.size() != 1)
+        continue;
+
+      if (isa<ReturnInst>(BB.getTerminator()))
+        continue;
+
+      if (MergeBlocks.count(&BB) != 0 || ContinueBlocks.count(&BB) != 0)
+        continue;
+
+      if (IsRequiredForPhiNode(&BB))
+        continue;
+
+      if (BB.getUniqueSuccessor() == nullptr)
+        continue;
+
+      BasicBlock *Successor = BB.getUniqueSuccessor();
+      std::vector<BasicBlock *> Predecessors(predecessors(&BB).begin(),
+                                             predecessors(&BB).end());
+      for (BasicBlock *Predecessor : Predecessors)
+        replaceBranchTargets(Predecessor, &BB, Successor);
+      ToRemove.push_back(&BB);
+    }
+
+    for (BasicBlock *BB : ToRemove)
+      BB->eraseFromParent();
+
+    return ToRemove.size() != 0;
+  }
+
+  bool addHeaderToRemainingDivergentDAG(Function &F) {
+    bool Modified = false;
+
+    auto MergeBlocks = getMergeBlocks(F);
+    auto ContinueBlocks = getContinueBlocks(F);
+    auto HeaderBlocks = getHeaderBlocks(F);
+
+    DomTreeBuilder::BBDomTree DT;
+    DomTreeBuilder::BBPostDomTree PDT;
+    PDT.recalculate(F);
+    DT.recalculate(F);
+
+    for (BasicBlock &BB : F) {
+      if (HeaderBlocks.count(&BB) != 0)
+        continue;
+      if (succ_size(&BB) < 2)
+        continue;
+
+      size_t CandidateEdges = 0;
+      for (BasicBlock *Successor : successors(&BB)) {
+        if (MergeBlocks.count(Successor) != 0 ||
+            ContinueBlocks.count(Successor) != 0)
+          continue;
+        if (HeaderBlocks.count(Successor) != 0)
+          continue;
+        CandidateEdges += 1;
+      }
+
+      if (CandidateEdges <= 1)
+        continue;
+
+      BasicBlock *Header = &BB;
+      BasicBlock *Merge = PDT.getNode(&BB)->getIDom()->getBlock();
+
+      bool HasBadBlock = false;
+      visit(*Header, [&](const BasicBlock *Node) {
+        if (DT.dominates(Header, Node))
+          return false;
+        if (PDT.dominates(Merge, Node))
+          return false;
+        if (Node == Header || Node == Merge)
+          return true;
+
+        HasBadBlock |= MergeBlocks.count(Node) != 0 ||
+                       ContinueBlocks.count(Node) != 0 ||
+                       HeaderBlocks.count(Node) != 0;
+        return !HasBadBlock;
+      });
+
+      if (HasBadBlock)
+        continue;
+
+      Modified = true;
+      Instruction *SplitInstruction = Merge->getTerminator();
+      if (isMergeInstruction(SplitInstruction->getPrevNode()))
```

Comment:
It would? If SplitInstruction is the first, `isMergeInstruction` would return false, meaning we'd call `splitBasicBlockBefore` on this instruction, which means creating an empty BB with just a branch to this instruction no?

Reply:
Yes, you are right, started overthinking this a bit. Marking as resolved.